protocols/floodsub/src/protocol.rs

```diff
@@ -135,13 +149,22 @@ where
     type Error = io::Error;
     type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
 
-    fn upgrade_outbound(self, mut socket: TSocket, _: Self::Info) -> Self::Future {
+    fn upgrade_outbound(self, socket: TSocket, _: Self::Info) -> Self::Future {
         Box::pin(async move {
             let bytes = self.into_bytes();
 
-            upgrade::write_length_prefixed(&mut socket, bytes).await?;
-            socket.close().await?;
-
+            let mut framed = Framed::new(
+                socket,
+                prost_codec::Codec::<std::vec::Vec<u8>>::new(MAX_MESSAGE_LEN_BYTES),
```

Comment:
Feeding bytes into this somewhat defeats the purpose of using `Framed` and `Codec`.

Instead, you'd want to change `self.into_bytes` to return you the `rpc_proto::Rpc` type and feed that into `framed.send`.

Reply:
You can then also remove this type hint, the compiler will figure it out!

```suggestion
                prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
```

Reply:
Hmmm, I kept getting errors about the type after I removed it. I looked elsewhere in the code and [saw in indentify ](https://github.com/libp2p/rust-libp2p/blob/c1e68b75c0729102ee43e2ec3cedc643b4d1f324/protocols/identify/src/protocol.rs#L181-L190)there was a section of code using a similar pattern of adding a type hint when created its `FramedWrite`.

Also in that code is the calls to the `send()` and `write()` functions, but for some reason that code does not generate warning about unsued returns like my code does :thinking: I did see that `protocols/gossipsub/src/config.rs` uses this pattern, but there it makes sense: the code is just setting fields and doesn't care about the result. For my code, I don't know why the compiler suddenly care about the return of those 2 calls.

Reply:
Right, my bad! It is needed :)